** 🧠 Day 3 – SQL via Python: NYC School Data Exploration **

In [51]:
import pandas as pd
import psycopg2

In [78]:
# DB connection setup using hardcoded credentials (for onboarding only)
conn = psycopg2.connect(
    dbname="neondb",
    user="neondb_owner",
    password="npg_CeS9fJg2azZD",
    host="ep-falling-glitter-a5m0j5gk-pooler.us-east-2.aws.neon.tech",
    port="5432",
    sslmode="require"
)
cur = conn.cursor()

In [ ]:
# Preview the first 5 rows from the high school directory table
query = "SELECT * FROM nyc_schools.high_school_directory LIMIT 5;"
df = pd.read_sql(query, conn)
df.head()

/var/folders/s3/jsrv8tv51h13z41rhcd_0n780000gn/T/ipykernel_7229/1731849884.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, conn)


dbn                                        school_name    borough building_code  phone_number    fax_number  grade_span_min  grade_span_max expgrade_span_min expgrade_span_max start_time  \
0  27Q260          Frederick Douglass Academy VI High School     Queens          Q465  718-471-2154  718-471-2890             9.0              12              None              None    7:45 AM   
1  21K559        Life Academy High School for Film and Music   Brooklyn          K400  718-333-7750  718-333-7775             9.0              12              None              None    8:15 AM   
2  16K393     Frederick Douglass Academy IV Secondary School   Brooklyn          K026  718-574-2820  718-574-2821             9.0              12              None              None    8:00 AM   
3  08X305                               Pablo Neruda Academy      Bronx          X450  718-824-1682  718-824-1663             9.0              12              None              None    8:00 AM   
4  03M485  Fiorello H. LaGuardia High School of Music & A...  Manhattan          M485  212-496-0700  212-724-5748             9.0              12              None              None    8:00 AM   

  end_time                                         priority01                                         priority02                                         priority03  \
0  2:05 PM  Priority to Queens students or residents who a...  Then to New York City residents who attend an ...               Then to Queens students or residents   
1  3:00 PM  Priority to New York City residents who attend...                    Then to New York City residents                                                      
2  2:20 PM                 Priority to continuing 8th graders  Then to Brooklyn students or residents who att...  Then to New York City residents who attend an ...   
3  3:50 PM  Priority to Bronx students or residents who at...  Then to New York City residents who attend an ...                Then to Bronx students or residents   
4  4:00 PM                    Open to New York City residents  Admission is based on the outcome of a competi...  Students must audition for each program (studi...   

                                          priority04                       priority05 priority06 priority07 priority08 priority09 priority10 location phone_number2 school_email  \
0                    Then to New York City residents                                                                                             None          None         None   
1                                                                                                                                                None          None         None   
2             Then to Brooklyn students or residents  Then to New York City residents                                                            None          None         None   
3                    Then to New York City residents                                                                                             None          None         None   
4  Students must be residents of New York City at...                                                                                             None          None         None   

                                        website                                             subway                                              bus grades2018 finalgrades  total_students  \
0  http://schools.nyc.gov/schoolportals/27/Q260                       A to Beach 25th St-Wavecrest                                        Q113, Q22       None        None             412   
1  http://schools.nyc.gov/schoolportals/21/K559  D to 25th Ave ; N to Ave U ; N to Gravesend - ...                         B1, B3, B4, B6, B64, B82       None        None             260   
2  http://schools.nyc.gov/schoolportals/16/K393             J to Kosciusko St ; M, Z to Myrtle Ave                B15, B38, B46, B47, B52, B54, Q24       None        None     

**🧮 1. School Distribution — number of schools per borough**

In [54]:
query = """
SELECT borough, COUNT(*) AS num_schools
FROM nyc_schools.high_school_directory
GROUP BY borough
ORDER BY num_schools DESC;
"""
df_schools = pd.read_sql(query, conn)
df_schools

/var/folders/s3/jsrv8tv51h13z41rhcd_0n780000gn/T/ipykernel_7229/2254687689.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_schools = pd.read_sql(query, conn)


borough  num_schools
0       Brooklyn          121
1          Bronx          118
2      Manhattan          106
3         Queens           80
4  Staten Island           10

**🎓 2. Language Learners — average % of ELL students per borough**

In [ ]:
# Preview the first 5 rows from the high school_demographics table
query = """
SELECT * 
FROM nyc_schools.school_demographics 
LIMIT 5
;
"""
df_columns_school_demographics = pd.read_sql(query, conn)
df_columns_school_demographics.head()

/var/folders/s3/jsrv8tv51h13z41rhcd_0n780000gn/T/ipykernel_7229/2965262378.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_columns_school_demographics = pd.read_sql(query, conn)


dbn                       Name  schoolyear fl_percent  frl_percent  total_enrollment prek   k grade1 grade2 grade3 grade4 grade5 grade6 grade7 grade8 grade9 grade10 grade11 grade12  ell_num  \
0  01M015  P.S. 015 ROBERTO CLEMENTE    20052006       89.4          NaN               281   15  36     40     33     38     52     29     38                                                    36   
1  01M015  P.S. 015 ROBERTO CLEMENTE    20062007       89.4          NaN               243   15  29     39     38     34     42     46                                                           38   
2  01M015  P.S. 015 ROBERTO CLEMENTE    20072008       89.4          NaN               261   18  43     39     36     38     47     40                                                           52   
3  01M015  P.S. 015 ROBERTO CLEMENTE    20082009       89.4          NaN               252   17  37     44     32     34     39     49                                                           48   
4  01M015  P.S. 015 ROBERTO CLEMENTE    20092010                    96.5               208   16  40     28     32     30     24     38                                                           40   

   ell_percent  sped_num  sped_percent  ctt_num  selfcontained_num  asian_num  asian_per  black_num  black_per  hispanic_num  hispanic_per  white_num  white_per  male_num  male_per  female_num  \
0         12.8        57          20.3       25                  9         10        3.6         74       26.3           189          67.3          5        1.8       158      56.2         123   
1         15.6        55          22.6       19                 15         18        7.4         68       28.0           153          63.0          4        1.6       140      57.6         103   
2         19.9        60          23.0       20                 14         16        6.1         77       29.5           157          60.2          7        2.7       143      54.8         118   
3         19.0        62          24.6       21                 17         16        6.3         75       29.8           149          59.1          7        2.8       149      59.1         103   
4         19.2        46          22.1       14                 14         16        7.7         67       32.2           118          56.7          6        2.9       124      59.6          84   

   female_per  
0        43.8  
1        42.4  
2        45.2  
3        40.9  
4        40.4

In [ ]:
# Calculate the average % of English Language Learners (ELL) per borough
query = """
SELECT hsd.borough, 
AVG(sd.ell_percent) AS avg_ell_percent
FROM nyc_schools.high_school_directory hsd
LEFT JOIN nyc_schools.school_demographics sd
ON sd.dbn = hsd.dbn
GROUP BY hsd.borough
ORDER BY avg_ell_percent DESC;
"""
df_avg_ell_percent = pd.read_sql(query, conn)
df_avg_ell_percent

/var/folders/s3/jsrv8tv51h13z41rhcd_0n780000gn/T/ipykernel_7229/2184604158.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_avg_ell_percent = pd.read_sql(query, conn)


borough  avg_ell_percent
0       Brooklyn              NaN
1         Queens              NaN
2  Staten Island              NaN
3          Bronx              NaN
4      Manhattan           7.5725

In [ ]:
# Query to list all unique school identifiers (DBN) 
# from the school_demographics table
query = """
SELECT DISTINCT(dbn)
FROM nyc_schools.school_demographics;
"""

df_dbn_school_demographics=pd.read_sql(query, conn)
df_dbn_school_demographics

/var/folders/s3/jsrv8tv51h13z41rhcd_0n780000gn/T/ipykernel_7229/2948907577.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_dbn_school_demographics=pd.read_sql(query, conn)


dbn
0   01M458
1   01M332
2   01M137
3   01M378
4   01M020
5   01M188
6   01M142
7   01M034
8   01M019
9   01M696
10  01M363
11  01M509
12  01M448
13  01M650
14  01M184
15  01M345
16  01M450
17  01M515
18  01M134
19  01M315
20  01M539
21  01M292
22  01M056
23  01M063
24  02M298
25  01M364
26  01M361
27  01M140
28  01M015
29  01M064
30  01M110
31  01M301

The result shows only Manhattan DBNs → meaning  school_demographics table in this database has demographic records only for Manhattan schools.

In [76]:
##Top 3 schools per borough by sped_percent
query= """
SELECT borough, school_name, sped_percent
FROM (
    SELECT hsd.borough,
           hsd.school_name,
           sd.sped_percent,
           ROW_NUMBER() OVER (
               PARTITION BY hsd.borough
               ORDER BY sd.sped_percent DESC
           ) AS rn
    FROM nyc_schools.high_school_directory hsd
    JOIN nyc_schools.school_demographics sd
      ON sd.dbn = hsd.dbn
) t
WHERE rn <= 3
ORDER BY borough, sped_percent DESC;
"""
df_top3_sped_percent = pd.read_sql(query, conn)
df_top3_sped_percent


/var/folders/s3/jsrv8tv51h13z41rhcd_0n780000gn/T/ipykernel_7229/1410962291.py:19: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_top3_sped_percent = pd.read_sql(query, conn)


borough                 school_name  sped_percent
0  Manhattan  East Side Community School          28.8
1  Manhattan  East Side Community School          27.7
2  Manhattan  East Side Community School          26.7

# 🧠 NYC School Data Exploration – Summary

## 🏫 School Distribution
- **Brooklyn**: 121 schools  
- **Bronx**: 118 schools  
- **Manhattan**: 106 schools  
- **Queens**: 80 schools  
- **Staten Island**: 10 schools  

➡️ Schools are fairly evenly distributed across the larger boroughs, with Staten Island having the fewest.

---

## 🎓 English Language Learners (ELL)
- Only **Manhattan** has valid ELL data: **~7.57% average**  
- All other boroughs show **NaN (no data)**  

⚠️ This indicates the `school_demographics` table in the training database contains records only for **Manhattan schools**.

---

## 🔗 Special Education (SPED)
- Current query output shows **repeated values for the same school** (East Side Community School in Manhattan).  
- This happens because `school_demographics` has **multiple rows per school** (likely for different years).  


---

## ✅ Key Insights
1. **School counts** by borough are complete from `high_school_directory`.  
2. **ELL data** is incomplete (only Manhattan is covered).  
3. **SPED analysis** requires collapsing duplicates per school before ranking.  
